lib/typescript-extensions.js

```diff
@@ -0,0 +1,169 @@
+module.exports = {
+    rules: {
+        /*
+            The following are extension rules that replace core JavaScript rules to support
+            TypeScript.
+            * When upgrading, changes to these rules can be identified in the typescript-eslint
+              changelog under features and breaking changes:
+              https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin#extension-rules
+            * In addition, the `npm run dev-print-typescript-props` command can be used to list
+              the expected extension properties.
+            * The value of the extension properties should match the value chosen by the
+              JavaScript / Airbnb configuration.
+        */
+
+        // Defined by Airbnb
+        'brace-style': 'off',
+        '@typescript-eslint/brace-style': ['error', '1tbs', { allowSingleLine: true }],
+
+        // Defined by NI
+        'comma-dangle': 'off',
+        '@typescript-eslint/comma-dangle': ['error', 'only-multiline'],
+
+        // Defined by Airbnb
+        'comma-spacing': 'off',
+        '@typescript-eslint/comma-spacing': ['error', { before: false, after: true }],
+
+        // Defined by Airbnb
+        'default-param-last': 'off',
+        '@typescript-eslint/default-param-last': 'off',
+
+        // Defined by Airbnb
+        'func-call-spacing': 'off',
+        '@typescript-eslint/func-call-spacing': ['error', 'never'],
+
+        // Defined by NI
+        indent: 'off',
+        '@typescript-eslint/indent': ['error', 4],
+
+        // Defined by Airbnb
+        'init-declarations': 'off',
+        '@typescript-eslint/init-declarations': 'off',
+
+        // Defined by Airbnb
+        'keyword-spacing': 'off',
+        '@typescript-eslint/keyword-spacing': ['error', {
+            before: true,
+            after: true,
+            overrides: {
+                return: { after: true },
+                throw: { after: true },
+                case: { after: true }
+            }
+        }],
+
+        // Defined by Airbnb
+        'lines-between-class-members': 'off',
+        '@typescript-eslint/lines-between-class-members': ['error', 'always', { exceptAfterSingleLine: false }],
```

Comment:
We were seeing this rule applied on more than just functions, but we're confused about whether that's intended behavior. We should see if that's a bug or not by comparing the behavior to the JS version of the rule. I'll investigate further.

Reply:
The ESLint version of the rule only officially supports functions, not fields, because class fields aren't a "stage 4" JavaScript feature yet (see comment in [this issue](https://github.com/eslint/eslint/issues/13295)). That said, their implementation works for fields.

The typescript-eslint version of the rule doesn't explicitly add capabilities for fields in TypeScript; it just inherits the capability of the ESLint rule.

So we're in a weird limbo where the rule works for fields but neither package wants to take credit for it so I'm not sure where the docs stating this support should belong. I'm not planning to submit an issue or PR to either repo unless someone feels strongly that I should. The proposal above will continue to work well for our use case.

Reply:
Enforcing lines between methods is common, because methods tend to be multiline, complex, and yet filled with whitespace. Lines between methods enforce boundaries and consistency, and I place lines between single line methods as well for the latter reason. However, lines between fields is not as common, because they are generally equivalent to declaring variables. Fields tend to be a single line the extra whitespace superfluous, even noisy. It also prevents grouping.

This is probably why the TS rule doesn't address fields. The common desire is to enforce space between methods. After testing, exceptAfterSingleLine doesn't even enforce multiline fields.

Reply:
@rajsite 🚀
We also need to make the same change to the eslint config.